### PR TITLE
An attempt at #432

### PR DESCRIPTION
**Pull request by KarimJedda**
Alright this might be a little bit of a stretch, but I believe the issue being floating point shenanigans. Since this required quite a bit of trial and error, I'll share my solution. If it is something acceptable, I can refine the PR, lint the code a bit and add a few tests, otherwise I'd be happy to see how you will solve it. Let me know what you prefer.

Debugging the two pieces of code mentioned in #432 , gives significantly different coordinates for the vertices, as well as creating quite a few more as you pointed out. 

To solve this, I've done three things:
- Used a `cKDTree` for all the spatial data structure stuff
- Introduced a tolerance level at which it's appropriate to merge two vertices
- Introduced a precision factor for the floating point coordinates 

The gist of the change involved reworking `_mesh_shape`, into shape (pun intended). 

There might certainly be a better solution to this, all tests are green so I think there should be no regressions. This also solves #311 

Here's the beautiful object, at last:
Screenshot from 2023-12-23 23-51-42

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,7 +40,7 @@ Development install
 ```
 git clone https://github.com/gumyr/build123d.git
 cd build123d
-python3 -m pip install -e .
+python3 -m pip install -r requirements.txt
```
Comment (jdegenstein) [inline README.md:43]: Can you revert this change? I just want to keep this PR focused on one issue.

**Comment by jdegenstein** [comment]
Nice work!

Couple other things here: can you make sure to run through the black formatter? It looked to me like there were a few formatting differences before/after black. Not directly relevant to this PR, but we might want to add a section to the docs regarding contributions to set out our standard practices.

Also, I wonder what the appropriate tolerance for this is? Currently you have it set to 1e-5, but my guess is that might be too high. For the 0.91 degree test case in #432 could you provide the coordinates of the offending vertices that are removed? That might give us an idea if it should be e.g. 1e-5, 1e-6, or 1e-9. 

Lastly, it would be a good idea to add at least one test case of the former failing behavior in case it comes back.

**Comment by gumyr** [comment]
I've got more testing to do but I think you're correct in that the problem is precision.  Mesher already does the elimination of redundant vertices so all I had to add was the following:
image

**Comment by gumyr** [comment]
Thanks but the less complex change (cd834d1a16211e8b8613514deef7ecf8fbf791b7) has fixed the problem.

**Comment by KarimJedda** [comment]
Fantastic, I'm very glad it worked out. I agree that your solution is much better, thank you both for the review!
